### PR TITLE
PACA-560: Create table for message errors

### PR DESCRIPTION
**Pull request by KyRobbins**
**JIRA Ticket:**
[PACA-560](https://jira.cms.gov/browse/PACA-560)

**User Story or Bug Summary:**

---

### What Does This PR Do?

Creates a message errors table to be used with RDA data to capture messages that generate errors upon ingestion through the RPC pipeline job.  This PR must be merged prior to #1086

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__rda_dead_letter_table.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+CREATE TABLE rda.message_errors (
+    sequence_number bigint                      not null,
+    claim_id        varchar(25)                 not null,
+    claim_type      varchar(20)                 not null,
+    api_source      varchar(24)                 not null,
+    created_date    timestamp with time zone    not null,
+    updated_date    timestamp with time zone    not null,
+    errors          ${type.jsonb}               not null,
+    message         ${type.jsonb}               not null,
+    primary key (sequence_number, claim_id, claim_type)
```
Comment (ericbuckley) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__rda_dead_letter_table.sql:22]: Why is claim_id part of the primary key?  I believe sequence number plus claim_type should uniquely map back to any RDA API message.

**Review by ericbuckley — DISMISSED** [review]

**Review by ericbuckley — APPROVED** [review]

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * This migration adds a message_errors table for tracking messages that resulted in
+ * an error during the message ingestion process via the gRPC pipeline job.  This table will
+ * provide a means to store the instigating data so it can be reviewed and troubleshooted.
+ *
+ * Keeping messages that generated errors in a separate persistent table will prevent it
+ * from being lost due to subsequent claim updates that may clear the error, thus
+ * making it more challenging to resolve the original issue.
+ */
+
+${logic.hsql-only} CREATE TYPE jsonb AS longvarchar;
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql:11]: @christrotteradhoc I believe this one was previously defined.  It appears on lines 96 and 114 in the DatabaseSchemaManager class

[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql:11]: my question is whether any of that is needed if we are doing this. can't we just reference the jsonb type below without bothering with placeholders?

[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql:11]: or alternatively, if we have the placeholders why do we need this line?

[the remark this answers is not in the document]
Comment (ericbuckley) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql:11]: oh I see now, good catch.  I wonder if this was a previous work around that Kyle didn't remove after he added the new `type.jsonb`.  @brianburton do you see any reason why we can't remove line 11?

[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql:11]: I had used both in my original jsonb prototype and that just got carried through in this PR.

[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V68__rda_dead_letter_table.sql:11]: I have removed the `type.json` placeholders and all tests pass under hsqldb and here in github.

**Review by brianburton — APPROVED** [review]

**Review by ericbuckley — APPROVED** [review]

**Review by christrotteradhoc — APPROVED** [review]

**Pull request merged**